doc/architectural/central-data-structures.md

```diff
@@ -0,0 +1,99 @@
+# Central Data Structures
+
+The following is some light technical documentation of the major data structures
+represented in the input transformation to Intermediate Representation (IR) inside
+CBMC and the assorted CProver tools.
+
+## goto_modelt
+
+The `goto_modelt` is the main data structure that CBMC (and the other tools) use to
+represent GOTO-IR (the `GOTO` Intermediate Representation).
+
+A `goto_modelt` is effectively a pair, consisting of:
+
+* A list of GOTO functions,
+* A symbol table containing symbol references for the symbols contained in the GOTO functions.
+
+In pseudocode, the type looks this:
+
+```js
+type goto_modelt {
+  type goto_functionst = list<goto_functiont>
+  type symbol_tablet = map<identifier, symbolt>
+}
+```
+
+The abstract interface of `goto_modelt` is outlined in the file
+[`src/goto-programs/abstract_goto_model.h`](../../src/goto-programs/abstract_goto_model.h).
+
+## goto_functiont
+
+A `goto_functiont` is also a product type. It's designed to represent a function
+at the IR level, and effectively it's the following ADT (in pseudocode):
```

Comment:
"IR" not defined. Above the definition was only for "GOTO-IR".

"and effectively it's the following ..." -> "like the following ..."

Reply:
Intermediate representation (IR) is defined in line 4 of the same file.

Reply:
Not quite, you defined "GOTO-IR", not "IR". I propose change line 4:
"GOTO-IR (GOTO Intermediate Representation)" to something like "GOTO intermediate representation (IR) denoted GOTO-IR"

Reply:
> Why say IR when we could just say GOTO?

Because GOTO is a CBMC concept, and IR is a lot more likely to be a known concept to a computer scientist, so by virtue of connecting the two concepts, we are leveraging people's pre-existing understanding of a general concept to relate it to the internal one.

Reply:
You already explained that in the first paragraph. Repeating it again is redundant.